zenoh/src/net/routing/interceptor/access_control.rs

```diff
@@ -0,0 +1,433 @@
+            if self.declare_subscriber(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Declare(Declare {
+            body: DeclareBody::DeclareQueryable(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.declare_queryable(key_expr) == Permission::Deny {
+                return None;
+            }
+        }
+        Some(ctx)
+    }
+}
+
+impl InterceptorTrait for EgressAclEnforcer {
+    fn compute_keyexpr_cache(&self, key_expr: &KeyExpr<'_>) -> Option<Box<dyn Any + Send + Sync>> {
+        Some(Box::new(key_expr.to_string()))
+    }
+    fn intercept(
+        &self,
+        ctx: RoutingContext<NetworkMessage>,
+        cache: Option<&Box<dyn Any + Send + Sync>>,
+    ) -> Option<RoutingContext<NetworkMessage>> {
+        let key_expr = cache
+            .and_then(|i| match i.downcast_ref::<String>() {
+                Some(e) => Some(e.as_str()),
+                None => {
+                    log::debug!("[ACCESS LOG]: Cache content was not of type String");
+                    None
+                }
+            })
+            .or_else(|| ctx.full_expr())?;
+
+        if let NetworkBody::Push(Push {
+            payload: PushBody::Put(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.put(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Request(Request {
+            payload: RequestBody::Query(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.get(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Declare(Declare {
+            body: DeclareBody::DeclareSubscriber(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.declare_subscriber(key_expr) == Permission::Deny {
+                return None;
+            }
+        } else if let NetworkBody::Declare(Declare {
+            body: DeclareBody::DeclareQueryable(_),
+            ..
+        }) = &ctx.msg.body
+        {
+            if self.declare_queryable(key_expr) == Permission::Deny {
+                return None;
+            }
+        }
+        Some(ctx)
+    }
+}
+pub trait AclActionMethods {
+    fn policy_enforcer(&self) -> Arc<PolicyEnforcer>;
+    fn interface_list(&self) -> Vec<i32>;
+    fn zid(&self) -> ZenohId;
+    fn flow(&self) -> Flow;
+
+    fn put(&self, key_expr: &str) -> Permission {
+        let policy_enforcer = self.policy_enforcer();
+        let interface_list = self.interface_list();
+        let zid = self.zid();
+        let mut decision = policy_enforcer.default_permission;
+        for subject in &interface_list {
+            match policy_enforcer.policy_decision_point(
+                *subject,
+                self.flow(),
+                Action::Put,
+                key_expr,
+            ) {
+                Ok(Permission::Allow) => {
```

Comment:
What is our permission priority?
From this code it looks like if we have an explicit allow, then deny are ignored.
It's right? Is this what is expected? Or did I not fully understand how it works here?

Reply:
The code loops over all the interfaces through which the connection has been made and if the action is allowed on any one of the interfaces (by the decision of the `policy_decision_point`), then it will be allowed and loop will be exited. This is different from permission priority implemented in `policy_decision_point` where the implicit `deny` is always given top priority.